File: src/clients/python/engine/otel.py
"""OpenTelemetry SDK initialization for Python strategy clients.

Mirrors Go utils.SetupOTelSDK. Creates TracerProvider and MeterProvider
with OTLP HTTP exporters. Configuration via OTEL_* environment variables.
"""
import os
import logging

from opentelemetry import trace, metrics
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.semconv.resource import ResourceAttributes

logger = logging.getLogger(__name__)

_initialized = False
_shutdown_fn = None


def setup_otel(service_name: str = None, service_version: str = "1.0.0"):
    """Initialize OTel SDK with OTLP HTTP exporters.

    Reads OTEL_EXPORTER_OTLP_ENDPOINT and OTEL_SERVICE_NAME from env vars.
    Returns a shutdown callable for cleanup.

    Idempotent: calling multiple times returns the same shutdown callable
    without creating duplicate providers.
    """
    global _initialized, _shutdown_fn

    if _initialized:
        logger.warning("setup_otel() already called; returning existing shutdown callable")
        return _shutdown_fn

    if service_name is None:
        service_name = os.getenv("OTEL_SERVICE_NAME", "grodt-strategy")

    resource = Resource.create({
        ResourceAttributes.SERVICE_NAME: service_name,
        ResourceAttributes.SERVICE_VERSION: service_version,
    })

    # Traces
    trace_exporter = OTLPSpanExporter()  # reads OTEL_EXPORTER_OTLP_ENDPOINT
    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(BatchSpanProcessor(trace_exporter))
    trace.set_tracer_provider(tracer_provider)

    # Metrics
    metric_exporter = OTLPMetricExporter()  # reads OTEL_EXPORTER_OTLP_ENDPOINT
    metric_reader = PeriodicExportingMetricReader(metric_exporter)
    meter_provider = MeterProvider(resource=resource, metric_readers=[metric_reader])
    metrics.set_meter_provider(meter_provider)

    def shutdown():
        """Flush and shut down both providers."""
        tracer_provider.force_flush()
        tracer_provider.shutdown()
        meter_provider.force_flush()
        meter_provider.shutdown()

    _initialized = True
    _shutdown_fn = shutdown

    logger.info("OTel SDK initialized: service_name=%s", service_name)
    return shutdown

File: src/go/utils/otel.go
package utils

import (
	"context"
	"errors"
	"os"
	"time"

	"go.opentelemetry.io/contrib/instrumentation/runtime"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlplog/otlploghttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	otelglobal "go.opentelemetry.io/otel/log/global"
	"go.opentelemetry.io/otel/propagation"
	sdklog "go.opentelemetry.io/otel/sdk/log"
	"go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.25.0"
)

// SetupOTelSDK bootstraps the OpenTelemetry pipeline with OTLP HTTP exporters
// for both traces and metrics. It configures the global TracerProvider,
// MeterProvider, and TextMapPropagator. The returned shutdown function must
// be called for proper cleanup (e.g., via defer).
//
// Configuration is controlled via OTEL_* environment variables (e.g.,
// OTEL_EXPORTER_OTLP_ENDPOINT). No endpoints are hardcoded.
func SetupOTelSDK(ctx context.Context, serviceName, serviceVersion string) (shutdown func(context.Context) error, err error) {
	var shutdownFuncs []func(context.Context) error

	// shutdown calls cleanup functions registered via shutdownFuncs.
	// The errors from the calls are joined.
	// Each registered cleanup will be invoked once.
	shutdown = func(ctx context.Context) error {
		var err error
		for _, fn := range shutdownFuncs {
			err = errors.Join(err, fn(ctx))
		}
		shutdownFuncs = nil
		return err
	}

	// handleErr calls shutdown for cleanup and makes sure that all errors are returned.
	handleErr := func(inErr error) {
		err = errors.Join(inErr, shutdown(ctx))
	}

	// Set W3C TraceContext + Baggage propagation
	prop := propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	)
	otel.SetTextMapPropagator(prop)

	// Create resource with service name, version, and instance ID (hostname)
	hostname, _ := os.Hostname()
	if hostname == "" {
		hostname = "unknown"
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(serviceName),
			semconv.ServiceVersion(serviceVersion),
			semconv.ServiceInstanceID(hostname),
		),
	)
	if err != nil {
		return nil, err
	}

	// Set up trace exporter (OTLP HTTP, endpoint from OTEL_* env vars)
	traceExporter, err := otlptrace.New(ctx, otlptracehttp.NewClient())
	if err != nil {
		return nil, err
	}

	// Set up TracerProvider
	tracerProvider := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(traceExporter),
		sdktrace.WithResource(res),
	)
	shutdownFuncs = append(shutdownFuncs, tracerProvider.Shutdown)
	otel.SetTracerProvider(tracerProvider)

	// Set up metric exporter (OTLP HTTP, endpoint from OTEL_* env vars)
	metricExporter, err := otlpmetrichttp.New(ctx)
	if err != nil {
		handleErr(err)
		return
	}

	// Set up MeterProvider
	meterProvider := metric.NewMeterProvider(
		metric.WithReader(metric.NewPeriodicReader(metricExporter)),
		metric.WithResource(res),
	)
	shutdownFuncs = append(shutdownFuncs, meterProvider.Shutdown)
	otel.SetMeterProvider(meterProvider)

	// Set up log exporter (OTLP HTTP, endpoint from OTEL_* env vars)
	logExporter, err := otlploghttp.New(ctx)
	if err != nil {
		handleErr(err)
		return
	}

	// Set up LoggerProvider
	loggerProvider := sdklog.NewLoggerProvider(
		sdklog.WithProcessor(sdklog.NewBatchProcessor(logExporter)),
		sdklog.WithResource(res),
	)
	shutdownFuncs = append(shutdownFuncs, loggerProvider.Shutdown)
	otelglobal.SetLoggerProvider(loggerProvider)

	// Start Go runtime metrics collection
	err = runtime.Start(runtime.WithMinimumReadMemStatsInterval(time.Second))
	if err != nil {
		handleErr(err)
		return
	}

	return
}
